libraries/AP_EFI/AP_EFI_Serial_MS.cpp

```diff
@@ -36,144 +34,132 @@ void AP_EFI_Serial_MS::update()
         return;
     }
 
-    uint32_t now = AP_HAL::millis();
-
-    const uint32_t expected_bytes = 2 + (RT_LAST_OFFSET - RT_FIRST_OFFSET) + 4;
-    if (port->available() >= expected_bytes && read_incoming_realtime_data()) {
-        last_response_ms = now;
+    if (read_incoming_realtime_data()) {
         copy_to_frontend();
     }
 
-    if (port->available() == 0 || now - last_response_ms > 200) {
+    uint32_t now = AP_HAL::millis();
+
+    if (now - internal_state.last_updated_ms > 200) {
         port->discard_input();
         // Request an update from the realtime table (7).
         // The data we need start at offset 6 and ends at 129
         send_request(7, RT_FIRST_OFFSET, RT_LAST_OFFSET);
     }
 }
 
-bool AP_EFI_Serial_MS::read_incoming_realtime_data() 
+bool AP_EFI_Serial_MS::read_incoming_realtime_data()
 {
-    // Data is parsed directly from the buffer, otherwise we would need to allocate
-    // several hundred bytes for the entire realtime data table or request every
-    // value individiually
-    uint16_t message_length = 0;
+    const uint32_t available_bytes = port->available();
+    if (available_bytes == 0) {
+        return false;
+    }
 
-    // reset checksum before reading new data
-    checksum = 0;
-    
-    // Message length field begins the message (16 bits, excluded from CRC calculation)
-    // Message length value excludes the message length and CRC bytes 
-    message_length = port->read() << 8;
-    message_length += port->read();
-
-    if (message_length >= 256) {
-        // don't process invalid messages
-        // hal.console->printf("message_length: %u\n", message_length);
+    // the following is 2 length bytes, the number of bytes we request
+    // in the packet we send and 4 bytes of checksum:
+    constexpr uint32_t expected_bytes = 2 + (RT_LAST_OFFSET - RT_FIRST_OFFSET) + 4;
+
+    // we must have *exactly* what we want in the buffer:
+    if (available_bytes != expected_bytes) {
```

Comment:
tridge didn't like this bit; he's concerned that some MegaSquirt boards might return fewer bytes than we request.

If that's the case this entire parser structure probably falls in a heap.

Reply:
This worked before, but given the packet structure there is no header, and no identifier of the first field reported. If we don't get the full set of fields we requested, it's impossible to be sure that we actually are handling the response we asked for. We might instead be getting a totally different partial response to some other range. I think this is sadly a good check to have, and it's just a limitation of a bad protocol.